### PR TITLE
Experimental next generation build cache protocol

### PR DESCRIPTION
**Pull request by lptr**
Experimental implementation of a new approach to storing data in the build cache. The purpose of this PR is to evaluate the implementation against the current-gen build cache, run benchmarks and dogfood it on the Gradle build itself.

The experimental feature can be enabled via the `org.gradle.unsafe.cache.ng` system property. 

The next-gen implementation is a drop-in replacement behind the `BuildCacheController` interface. Aside from differences in performance, build behavior should be unchanged, i.e. the same tasks should be loaded/stored in the cache with either implementation.

The next-gen cache works with existing remote cache backends, though it is best used with `HttpBuildCacheService`. For local cache it uses an H2 database stored in the directory configured for the local cache. No change in configuration is required for existing builds.

Cache keys are intentionally different for next-gen to avoid collisions when using the same backend and switching between next-gen and current-gen.

## Working mechanism

In the current-gen build cache implementation the output of a unit of work (such as a task) is stored atomically in the build cache in a single cache artifact. The artifact is a gzip'd TAR file that contains both metadata about the work's outputs (which file goes where, how long did it take to create the outputs, the build invocation ID that created the artifact etc.) as well as the actual contents of the output files. The cache key used to address the outputs is generated from the normalized inputs of the work.

This approach has served us well in the past, but it has some downsides that we'd long wanted to improve upon. The main issue is the redundancy of the format, for different executions of a unit of work it is quite typical to have some output files change, while others will stay the same as before. Just think of a `JavaCompile` task compiling 100 source files: the developer will likely change a few of them at a time, which will result in some `.class` files changing on the output, but the majority will stay the same, yet our current-gen caching will store the unchanging parts as well as the changing ones every time.

In the next-gen cache individual output file contents are stored in a CAS (content-addressed storage), ensuring that each output is only stored once. The cache key (generated the same way from inputs) addresses a lightweight manifest that contains the metadata and references to the file contents stored in the CAS.

We expect the reduced redundancy to reduce both local cache sizes and the amount of data having to be transferred from remote caches. Furthermore, the separation of metadata and content allows us to implement some further improvements that would not be possible without it.

## Missing functionality

Some features are not implemented, or are not fully implemented:

- There is no cache cleanup, and the size of the database is allowed to increase indefinitely.
- Any error with the cache (like a corrupted entry or a network error) will make the build fail. Current-gen implementation has some cushions around these cases that wasn't yet migrated.
- Build scan support is partial, and some data about cache operations will be missing.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/build-cache/src/main/resources/db/migration/V001__init_schema.sql [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+CREATE TABLE IF NOT EXISTS filestore.catalog
 (
     id            BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
     entry_key     VARCHAR(32)         NOT NULL,
     entry_size    BIGINT              NOT NULL,
     entry_content BINARY LARGE OBJECT NOT NULL
 );
 
-CREATE INDEX idx_catalog_entry_key ON filestore.catalog (entry_key);
+CREATE INDEX IF NOT EXISTS idx_catalog_entry_key ON filestore.catalog (entry_key);
```
Comment (lptr) [inline subprojects/build-cache/src/main/resources/db/migration/V001__init_schema.sql:11]: Do we need this index now? Won't the primary key get indexed automatically?

Comment (asodja) [reply subprojects/build-cache/src/main/resources/db/migration/V001__init_schema.sql]: Check next commit, should fix that

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/build-cache/build.gradle.kts [hunk 2]
```diff
@@ -24,6 +24,13 @@ dependencies {
     implementation(libs.slf4jApi)
     implementation(libs.guava)
     implementation(libs.inject)
+    implementation("com.h2database:h2:2.1.214")
```
Comment (asodja) [inline subprojects/build-cache/build.gradle.kts:27]: We should move this to managed dependencies

File: subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheEntryWriter.java [hunk 3]
```diff
@@ -25,6 +29,14 @@
  * @since 3.3
  */
 public interface BuildCacheEntryWriter {
+    /**
+     * Opens a stream to read the contents to be written to cache.
+     *
+     * @throws IOException when an I/O error occurs while opening the stream.
+     */
+    @Incubating
+    InputStream openStream() throws IOException;
```
Comment (asodja) [inline subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheEntryWriter.java:38]: Is it safe we add this?

File: subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheEntryWriter.java [hunk 4]
```diff
@@ -33,7 +45,13 @@ public interface BuildCacheEntryWriter {
      * @param output output stream to write build cache entry to
      * @throws IOException when an I/O error occurs when writing the cache entry to the given output stream
      */
-    void writeTo(OutputStream output) throws IOException;
+    default void writeTo(OutputStream output) throws IOException {
```
Comment (asodja) [inline subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheEntryWriter.java:48]: Is it safe we change this to default?

**Review by asodja — CHANGES_REQUESTED** [review]
Concerned about API changes

File: gradle/verification-metadata.xml [hunk 5]
```diff
@@ -1190,6 +1196,11 @@
             <pgp value="7c669810892cbd3148fa92995b05ccde140c2876"/>
          </artifact>
       </component>
+      <component group="org.flywaydb" name="flyway-core" version="9.12.0">
```
Comment (asodja) [inline gradle/verification-metadata.xml:1199]: I remembered we can remove that now, we removed Flyway dependency

Comment (asodja) [reply gradle/verification-metadata.xml]: Removed

File: subprojects/build-cache/build.gradle.kts (hunk 2 above)
Comment (asodja) [reply subprojects/build-cache/build.gradle.kts]: fixed

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Review by asodja — APPROVED** [review]
LGTM

**Comment by lptr** [comment]
@bot-gradle test this.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Pull request merged**